### PR TITLE
[Navigator] add optional callbacks to routing functions

### PR DESCRIPTION
**Pull request by pwmckenna**

### CODE REVIEW

**Comment by ericvicenti** [comment]
This might be confusing because the callbacks will resolve before the animation completes in the case of `push`, but it will call back after the animation for `pop`, etc.

Could you use the willFocus/didFocus API for this?

**Comment by pwmckenna** [comment]
is the reason that it resolves before the animation completes for push because `_transitionTo` kicks something off (and doesn't accept a callback)? I could try to fix that. What the callbacks are good for (and where willFocus/didFocus would struggle i believe) is doing something like popping, waiting for that to finish, then pushing something else on. I could certainly write a wrapper around `willFocus`/`didFocus` that ties them to the most recent navigation and provides the same type of callback, but it would be much more complicated.

**Comment by ericvicenti** [comment]
Ok, I see what you mean. This API does sound useful, even though we haven't needed it internally yet.

If you look at `_transitionTo`, it actually does accept a callback. You could pass it through and make it call back after the `push` transition.

**Comment by pwmckenna** [comment]
fantastic! i'll patch it up with the transitionTo callback. I have no idea how to write tests for this. suggestions?

**Comment by ericvicenti** [comment]
Right now we're totally lacking tests for navigator, partially because it can be tricky to mock out the animations. I'm going to add some tests in the next week or two, and I can also add a few tests for the callbacks.

**Comment by ide** [comment]
I think we want willLoseFocus and didLoseFocus lifecycle events, or something of that nature. Worth looking at UIViewController's transition API (it's complex...) and Android fragments to get a sense of what events a scene might want to receive and when those events should be fired.

**Comment by pwmckenna** [comment]
@ericvicenti made the `push` change, and added cb arguments to the additional exposed methods that I missed initially. What is the process for getting things merged in for this project? Should I squash commits? Wait for the tests (which I'd be more than happy to help with btw)? Happy to help however.

@ide Those lifecycle hooks sound super useful. I'd be happy to help with those as well, though I don't have any experience with UIViewController or Android fragments, and I'm not super in tune with what the long term goals of this project are as they pertain to android.

**Comment by ericvicenti** [comment]
@facebook-github-bot import

I'll test this manually for now and get it merged this week. Meanwhile, if you want to get going on the tests, I can point you in the right direction tomorrow.

File: Libraries/CustomComponents/Navigator/Navigator.js [hunk 1]
```diff
@@ -1009,11 +1009,11 @@ var Navigator = React.createClass({
     return destIndex;
   },
 
-  _jumpN: function(n) {
+  _jumpN: function(n, cb) {
     var destIndex = this._getDestIndexWithinBounds(n);
     var requestTransitionAndResetUpdatingRange = () => {
       this._enableScene(destIndex);
-      this._transitionTo(destIndex);
+      this._transitionTo(destIndex, cb);
```
Comment (ericvicenti) [inline Libraries/CustomComponents/Navigator/Navigator.js]: `cb` is the fourth argument in `_transitionTo`, so I don't think this will work

Comment (pwmckenna) [inline Libraries/CustomComponents/Navigator/Navigator.js]: good call. updated.

**Comment by ericvicenti** [comment]
Sorry for the delay here. I've been talking with @ide and have had a change of heart: instead of having callbacks on all of these actions, we could add focus lifecycle events that get called on the scenes:

`willFocus`
`didFocus`
`willLooseFocus`
`didLooseFocus`

I think these lifecycle methods will be more versatile than callbacks on the actions. Thoughts?

**Comment by ide** [comment]
Here are the specifics as a strawman:

- `willFocus`: when the navigator has committed to displaying a scene. For example, if you pan from the edge of the screen far enough to dismiss the topmost scene, the navigator will fire `willFocus` for the scene below when you lift your finger because the navigator has committed to dismissing the top scene. If you start panning but then don't pan far enough, the navigator does not fire the `willFocus` event because it will not focus the scene below (sounds obvious... just stating it clearly).
- `didFocus`: when the navigator has finished completely displaying a scene. If there is a transition animation after `willFocus`, then the navigator fires `didFocus` after the transition completes. If there is no transition, then the navigator simply fires `didFocus` after `willFocus`.
- `willBlur`: fired for the dismissed scene right before the navigator fires `willFocus` for the scene it will transition to
- `didBlur`: fired for the dismissed scene right before the navigator fires `didFocus` for the scene that it did transition to

**Pull request closed without merging**